splice_si_api/src/main/java/com/splicemachine/si/impl/store/IgnoreTxnSupplier.java

```diff
@@ -0,0 +1,112 @@
+
+/*
+ * Copyright (c) 2012 - 2017 Splice Machine, Inc.
+ *
+ * This file is part of Splice Machine.
+ * Splice Machine is free software: you can redistribute it and/or modify it under the terms of the
+ * GNU Affero General Public License as published by the Free Software Foundation, either
+ * version 3, or (at your option) any later version.
+ * Splice Machine is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY;
+ * without even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU Affero General Public License for more details.
+ * You should have received a copy of the GNU Affero General Public License along with Splice Machine.
+ * If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package com.splicemachine.si.impl.store;
+
+import com.splicemachine.access.api.PartitionAdmin;
+import com.splicemachine.access.api.PartitionFactory;
+import com.splicemachine.access.configuration.HBaseConfiguration;
+import com.splicemachine.encoding.MultiFieldDecoder;
+import com.splicemachine.si.api.data.TxnOperationFactory;
+import com.splicemachine.storage.*;
+import com.splicemachine.utils.SpliceLogUtils;
+import org.apache.hadoop.hbase.util.Pair;
+
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+
+/**
+ * Created by jyuan on 10/30/17.
+ */
+public class IgnoreTxnSupplier {
+    private Set<Pair<Long, Long>> cache;
+    private EntryDecoder entryDecoder;
+    final private PartitionFactory partitionFactory;
+    final private TxnOperationFactory txnOperationFactory;
+    private boolean ignoreTxnTableExists;
+    private boolean initialized = false;
+
+    public IgnoreTxnSupplier(PartitionFactory partitionFactory, TxnOperationFactory txnOperationFactory) {
+        this.partitionFactory = partitionFactory;
+        this.txnOperationFactory = txnOperationFactory;
+        cache = new HashSet<>();
+    }
+
+    public boolean shouldIgnore(Long txnId) throws IOException{
+
+        boolean ignore = false;
+        if (!initialized) {
+            init();
+            initialized = true;
+        }
+        if (!cache.isEmpty()) {
+            for (Pair<Long, Long> range : cache) {
+                if (txnId > range.getFirst() && txnId < range.getSecond()) {
+                    ignore = true;
+                    break;
+                }
```

Comment:
My understanding is that the pairs will be in order in the cache. Would it be better to add 
else if (txnId <= range.getFirst()) {
     return ignore;
}
So that we can terminate early. I'm thinking if the system has gone through restore quite a few times, maybe this could help to reduce some performance impact?

Reply:
Building on top of that, I think at some point it was discussed having a global min/max pair to avoid checking the cache in the common case. For instance, if the pairs are (10, 20), (100, 120), the global min/max would be (10, 120). When we check txn 5 or 200, since they're outside the min/max we don't have to traverse the cache.